ui/src/variables/utils/hydrateVars.ts

```diff
@@ -0,0 +1,381 @@
+    const cyclicPath = findCyclicPath(node)
+
+    if (cyclicPath) {
+      for (const invalidNode of cyclicPath) {
+        invalidNode.status = RemoteDataState.Error
+      }
+    }
+  }
+}
+
+/*
+  Given a node, mark all ancestors of that node as `Error`.
+
+  A node `a` is an ancestor of `b` if there exists a path from `a` to `b`.
+
+  Traversing the ancestors of a node stops at `Error` nodes. In this case it is
+  assumed that every ancestor of the `Error` node has also been marked as
+  `Error`; this is to avoid getting caught in cycles. Nodes within cycles and
+  nodes leading to cycles should have already been invalidated at this point by
+  `invalidateCycles`.
+*/
+const invalidateAncestors = (node: VariableNode): void => {
+  for (const parent of node.parents) {
+    if (parent.status !== RemoteDataState.Error) {
+      parent.status = RemoteDataState.Error
+      invalidateAncestors(parent)
+    }
+  }
+}
+
+const extractResult = (graph: VariableNode[]): VariableValuesByID => {
+  const result = {}
+
+  for (const node of graph) {
+    if (node.status === RemoteDataState.Error) {
+      node.values = errorVariableValues()
+    }
+
+    result[node.variable.id] = node.values
+  }
+
+  return result
+}
+
+/*
+  Given a list of `variables`, execute their queries to retrieve the possible
+  values for each variable.
+
+  Since variables can make use of other variables, this process is more
+  complicated then simply executing every query found in a variable. Instead,
+  we:
+
+  1. Construct a graph that represents the dependency structure between
+     variables. A variable `a` depends on variable `b` if the query for `a`
+     uses the variable `b`.
+     
+     Each node in the graph has a hydration status: it it either `NotStarted`,
+     `Loading` (values are being fetched), `Error` (values failed to fetch or
+     cannot be fetched), or `Done` (values have been fetched successfully).
+     When the graph is constructed, all nodes a marked as `NotStarted`.
+
+  2. Find all cycles in the graph and mark nodes within a cycle and nodes that
+     lead to a cycle as `Error`. We cannot resolve variables with cyclical
+     dependencies.
+
+  3. Find the leaves of the graph (the nodes with no dependencies) and begin to
+     load their values.
+
+     a. If loading the values succeeds, we mark that node as `Done` and check
+        if every child of the parent of this node is `Done`. If this is the
+        case, we start loading the parent.
+
+     b. If loading the values fails, we mark that node and every one of its
+        ancestors as as `Error`.
+
+     c. If the parent node is `Error`, we stop resolving along this path. Steps
+        2 and 3b guarantee that every ancestor of this `Error` node is also
+        `Error`.
+
+  4. By the time all node resolution has succeeded or failed, every node in the
+     graph has now either been resolved with values or marked with `Error`.
+
+*/
+export const hydrateVars = (
```

Comment:
This is a good place to start reading in this file. All other functions support this function.